README.rst
README
======

**Twerp is the telephone hackers toolkit.**

Twerp is also:

 * A command-line interface for the Twilio API

 * Your own Phone Company. From the command-line!

 * A tool that will revolutionize crank calling (only legal from California to the French parts of Canada. Note: IANALBISEEOC (I am not a lawyer but I've seen every episode of Cops).

Features:

 * Buy phone numbers from the command-line in the U.S., toll free and wherever Twilio sells them
 * Manage Twilio accounts: create new sub-accounts, list, rename accouts
 * Make phone calls from the command-line 
 * Conrtol call flow from the command-line using stateless TwiML transactions (no web app necessary)
 * Command-line driven conference calls
 * Send and receive SMS text messages
 * Read your Twilio logs from the command-line
 * Modify the flow of calls or conferences in progress with a curses based comand-line interface
 * Do lots of stuff without going to your dashboard on the twilio.com website


TODO:

 * Plugin system based on Python entry_points
 * Plugin to launch Bottle web app and localtunnel.com it!
 * Plugin for Phox Flask webapp
 * Create API for using alternate interfaces such as urwid to make complete SMS or phone app GUIs.
 * Make an entire website about what I could do with Twilio+Twerp.


.. contents::


Installation
------------

pip install twerp



Configure twerp
---------------

~/.twerprc


ACCOUNT_SID=a902830980980980ff987yada

AUTH_TOKEN=98798asdfas9df87sadf987yada

CALLERID=+12135551212



Usage::
=======

    Usage: twerp [options]

    Options:
      -h, --help            show this help message and exit
      --version             Show twerp version and exit.
      -v, --verbose         Show more output stuff.
      --debug               Show debugging information.
      -q, --quiet           Show less output.

      Common options:
        These options can be used for both SMS and voice calls.

        -c CALLERID, --callerid=CALLERID
                            Phone number you are calling from or texting from.
        -i, --interactive   Go into interactive command-line mode after dialing
                            (voice or conferences).

      Voice call options:
        Place phone calls, execute TWIML.

        -d +12135551212,+14155551212, --dial=+12135551212,+14155551212
                            List of numbers to dial, comma-separated.
        -y Say something., --say=Say something.
                            Use with --dial to say something.
        -u URL of TWIML, --url=URL of TWIML
                            URL of TWIML to pass call with --call
        -b +12135551212, --buy=+12135551212
                            Buy a specific phone number listed with -x or -a
        -a AREA CODE, --area-code=AREA CODE
                            Search for phone number to purchase by area code. Use
                            -b to purchase from these results.

        -x CONTAINS, --contains=CONTAINS
                            Search for phone number to purchase by numbers or
                            letters it contains.

      Conference (voice) options:
        These options can be used for voice conference calls.

        -f +12135551212,+14155551212, --conference=+12135551212,+14155551212
                            Start conference with list of numbers to dial, comma-
                            separated.

        -o ROOM, --room=ROOM
                            Room to join for voice conference.
        -e, --conferences   Show conferences in-progress.
        -p, --conference-participants
                            Show participants for all conferences in-progress.

      SMS options:
        Send and reveive SMS text messages.

        -m <TXT MSG>, --message=<TXT MSG>
                            Send SMS text message

        -s +12135551212,+14155551212, --sms=+12135551212,+14155551212
                            Send SMS text message to list of numbers.

        -l, --list-sms      Show incoming SMS messages.

      Reporting options:
        List your Twilio phone numbers and information about each.

        -n, --notifications
                            Show notifications from Twilio API (error messages and
                            warnings).
        -r, --numbers       Show all my Twilio phone numbers. Use -Nv for detailed
                            info on each number.
        --sid=SID           Show log for given SID

      Applications:
        Twilio Application information.

        --applications      Show all my Twilio Applications.


      Accounts:
        Twilio account and sub-account management

        --list-accounts
            List all Twilio accounts and sub-accounts.

        --create-sub-account=NAME
            Create sub-account named 'NAME'

        --rename-sub-account=NAME
            Rename account or sub-account using 'NAME'

Interactive Mode
================


The Prompt
----------

The prompt will have part of the SID if a call is in progress:

    twerp (CA3abc...) >>

If you hang up a call, for example, there will be no SID, so the prompt will look like this:

    twerp (...) >>


Interactive Mode Commands
-------------------------

 * list - List all calls in progress, ringing or queued
 * hangup - Hang up call associated with SID shown in prompt
 * nuke - Hang up all calls associated with account. ALL OF THEM!
 * forward <nnnnnnnnnn> - Redirect current call to another phone number
 * url <URL> - Redirect flow of call to TwiML at a URL
 * info [<SID>] - Show info for current SID or SID given
 * sid <SID> - Change the current SID associated with interactive-mode


Installation
------------
pip install twerp


Development
-----------

 * We're on Freenode #twerp
 * We're on Twitter @TwerpForTwilio
 * We're on Github https://github.com/cakebread/twerp
 * We're on continuous integration a la Jenkins http://cakebread.info:8080
 * Fork it, phone it.

See http://blog.cakebread.info/



NEWS.txt
.. Twerp News

News
====

0.0.5
-----

*Release date: 4/4/2012

* Added conference cabability
* Added account management, list, rename and create new Twilio sub-accounts
* Added application management: list all Twilio apps
* Added feature to buy new phone numbers, basic searching by area code or phrase
* Renamed a lot of command-line options
* More unit tests


0.0.3
-----

*Release date: 1/17/2012

* Added --interactive command-line feture to control calls in progress
* Changed many keyword options, we're still super++ pre-alpha, so I can do that.
* Added docs at http://twerp.rtfd.org/


0.0.2
-----

*Release date: 1/6/2012

* Added voice dialing

0.0.1
-----

*Release date: 1/5/2012

* Initial release





tests/test_twiliolib.py


from httplib2 import ServerNotFoundError
from mock import Mock, MagicMock, patch
from twerp.twilio_support import RestClient


def test_send_sms():

    real = RestClient()
    #Normal call
    real.send_sms = Mock()
    real.send_sms(['+13233333333'], 'message', False, '+18183333333')


    


def test_send_sms_server_not_found():

    real = RestClient()
    real.send_sms = Mock(side_effect=ServerNotFoundError('Server not found, dude.')).return_value = 2



    




src/twerp/mytwilio.py

"""

mytwilio.py
======

Desc: Send and receive SMS or make calls via Twilio REST API

Author: Rob Cakebread <cakebread @ gmail*com>

License : BSD

"""

__docformat__ = 'restructuredtext'


import os
import sys
import logging
from configobj import ConfigObj
from urllib import quote_plus

from httplib2 import ServerNotFoundError
from twilio.rest import TwilioRestClient
from twilio import TwilioRestException
from twilio.rest.resources import Call
from clint.textui import colored, puts, indent

logger = logging.getLogger("twerp")
twerp_config = os.path.join(os.path.expanduser("~"), '.twerprc')

if not os.path.exists(twerp_config):
    print """Create ~/.twerprc with:

    AUTH_TOKEN = lkasjdfkljsl
    ACCOUNT_SID = iowjroiwjeroij
    CALLER_ID = +13235551212
    """
    sys.exit(1)
config = ConfigObj(twerp_config)

AUTH_TOKEN = config['AUTH_TOKEN']
ACCOUNT_SID = config['ACCOUNT_SID']
CALLER_ID = config['CALLER_ID']
MAX_LENGTH = 160

'''
These are for showing detailed info about your Twilio phone numbers. It uses
TwilioRestClient.phone_numbers. There are some I didn't cover because they
require making multiple API calls, I don't know what they do yet, or aren't
applicable.
'''
NUMBER_IDS = ['account_sid', 'api_version', 'auth', 'base_uri', 'capabilities',
            'date_created', 'date_updated', 'friendly_name', 'id_key',  'name',
            'phone_number', 'sid', 'sms_application_sid',
            'sms_fallback_method', 'sms_fallback_url', 'sms_method', 'sms_url',
            'status_callback', 'status_callback_method', 'subresources', 'uri',
            'voice_application_sid', 'voice_caller_id_lookup',
            'voice_fallback_method', 'voice_fallback_url', 'voice_method',
            'voice_url']


class FakeCall(object):
    '''For making --pretend calls'''
    def __init__(self):
        self.sid = "YOURSID"
        self.status = "in progress(test)"


def trim(text_msg):
    '''Text messages have a finite length'''
    return text_msg[0:MAX_LENGTH]


def notifications(verbose=False):
    '''Print error messages and warnings from Twilio'''
    client = TwilioRestClient(ACCOUNT_SID, AUTH_TOKEN)
    notes = client.notifications.list()
    notes.reverse()

    for notification in notes:
        print "=="
        print notification.date_created
        print notification.call_sid
        print "Error code: %s" % notification.error_code
        print "Error msg: %s" % notification.message_text
        print "Request URL: %s" % notification.request_url
        print "Error: %s" % notification.log
        if verbose:
            print notification.more_info
        print


def sid_call(sid):
    """Print results for given SID"""

    client = TwilioRestClient(ACCOUNT_SID, AUTH_TOKEN)

    try:
        call = client.calls.get(sid)
    except ServerNotFoundError, e:
        logger.error(e)
        return 1
    except TwilioRestException, e:
        print e
        return 1

    print 'date: %s' % call.date_created
    print 'from: %s' % call.from_
    print 'to: %s' % call.to
    print 'status: %s' % call.status
    print 'direction: %s' % call.direction
    print


def get_sms_sid(sid):
    """Print results for given SID"""

    client = TwilioRestClient(ACCOUNT_SID, AUTH_TOKEN)

    try:
        sms = client.sms.messages.get(sid)
    except ServerNotFoundError, e:
        logger.error(e)
        return 1
    except TwilioRestException, e:
        print e
        return 1

    print 'date: %s' % sms.date_sent
    print 'from: %s' % sms.from_
    print 'to: %s' % sms.to
    print 'status: %s' % sms.status
    print 'body: %s' % sms.body
    print


def hangup(sid):
    '''Hangup a call by SID'''
    client = TwilioRestClient(ACCOUNT_SID, AUTH_TOKEN)

    try:
        client.calls.hangup(sid)
        print "Call hung up."
        print
    except TwilioRestException, e:
        logger.error(e)
        return 1


def call_url(sid, url):
    '''Call a URL/Twimlet'''
    client = TwilioRestClient(ACCOUNT_SID, AUTH_TOKEN)

    try:
        return client.calls.route(sid, url, method="POST")
    except TwilioRestException, e:
        logger.error(e)
        return 1


def hangup_all_calls():
    '''Call a URL/Twimlet'''
    client = TwilioRestClient(ACCOUNT_SID, AUTH_TOKEN)
    calls = client.calls.list(status=Call.IN_PROGRESS)
    for c in calls:
        print "Hung up IN_PROGRESS SID: %s  From:%s" % (c.sid, c.from_)
        c.hangup()
    calls = client.calls.list(status=Call.RINGING)
    for c in calls:
        print "Hung up RINGING SID: %s  From:%s" % (c.sid, c.from_)
        c.hangup()

    calls = client.calls.list(status=Call.QUEUED)
    for c in calls:
        print "Hung up QUEUED SID: %s  From:%s" % (c.sid, c.from_)
        c.hangup()
    print


def list_calls():
    '''List calls IN_PROGRESS, RINGING, or QUEUED'''
    client = TwilioRestClient(ACCOUNT_SID, AUTH_TOKEN)
    calls = client.calls.list(status=Call.IN_PROGRESS)
    for c in calls:
        print "SID: %s" % c.sid
        print "From: %s" % c.from_
        #print "Answered by: %s" % c.answered_by
        #This always shows None, maybe we can't get till call ends
        print "Status: %s" % c.status
        print "Direction: %s" % c.direction
        print "Start: %s" % c.start_time

    calls = client.calls.list(status=Call.RINGING)
    for c in calls:
        print "Ringing: %s" % c.sid

    calls = client.calls.list(status=Call.QUEUED)
    for c in calls:
        print "Queued: %s" % c.sid


def call_numbers(recipients, verbose=False, callerid=CALLER_ID, url=None,
        say=None):
    """
    callerid: string
    recipients: list of strings representing each phone number
    """

    if callerid is None:
        callerid = CALLER_ID

    client = TwilioRestClient(ACCOUNT_SID, AUTH_TOKEN)
    if say:
        #Let's hope this URL never goes away.
        url = '''http://twimlets.com/message?Message%5B0%5D='''
        text = quote_plus(say)
        url = '%s%s' % (url, text)
    for phone in recipients:
        logger.info("Placing call to: %s" % phone)
        try:
            call = client.calls.create(to=phone, from_=callerid, url=url)
        except ServerNotFoundError, e:
            logger.error(e)
            return None
        except TwilioRestException, e:
            logger.error(e)
            return None

        #call = FakeCall()
        logger.info("Status: %s" % call.status)
        logger.info("SID: %s" % call.sid)
    return call.sid


def send_sms(recipients, message, verbose=False, callerid=CALLER_ID):
    """
    callerid: string
    recipients: list of strings representing each phone number
    message: Text to send via SMS"""

    client = TwilioRestClient(ACCOUNT_SID, AUTH_TOKEN)

    for phone in recipients:
        try:
            message = client.sms.messages.create(to=phone, from_=callerid,
                    body=message)
        except ServerNotFoundError, e:
            logger.error(e)
            return 1
        except TwilioRestException, e:
            logger.error(e)
            return 1
        if verbose:
            logger.info("Status: %s" % message.status)
            logger.info("SID: %s" % message.sid)
            logger.info("From: %s" % message.from_)


def list_numbers(verbose=False):
    """List all my Twilio numbers"""
    client = TwilioRestClient(ACCOUNT_SID, AUTH_TOKEN)
    try:
        for number in client.phone_numbers.iter():
            puts(number.friendly_name)
            if verbose:
                with indent(4, quote=colored.blue(' >')):
                    for id in NUMBER_IDS:
                        puts("%s: %s" % (colored.green(id),
                            getattr(number, NUMBER_IDS[NUMBER_IDS.index(id)])))

    except ServerNotFoundError, e:
        logger.error(e)
        return 1


def list_sms():
    """
    Retrieve list of all sms messages

    TODO: Add filtering for from, to and date
    """

    #from_ = CALLER_ID
    client = TwilioRestClient(ACCOUNT_SID, AUTH_TOKEN)
    try:
        messages = client.sms.messages.list()
    except ServerNotFoundError, e:
        logger.error(e)
        return 1
    except TwilioRestException, e:
        print e
        return 1
    messages.reverse()
    for message in messages:
        print "%s t:%s f:%s sid:%s\n\t%s" % (message.date_sent, message.to,
            message.from_, message.sid, message.body)


src/twerp/cli.py
# PPylint: disable-msg=W0613,W0612,W0212,W0511,R0912,C0322,W0704
# W0511 = XXX (my own todo's)

"""

cli.py
======

Desc: Command-line module

Author: Rob Cakebread <cakebread a t gmail.com>

License : BSD

"""

__docformat__ = 'restructuredtext'


import cmd
import sys
import optparse
import logging
from urllib import quote_plus

from clint.textui import colored

from twerp.__init__ import __version__ as VERSION
from twerp.twilio_support import RestClient


class Interactive(cmd.Cmd):
    """Cmmand processor"""
    def __init__(self, client, sid=None):
        cmd.Cmd.__init__(self)
        self.sid = sid
        self.logger = logging.getLogger("twerp")
        self.client = client

    def emptyline(self):
        return

    def postloop(self):
        '''Print a blank line after program exits interactive mode'''
        print '\n'

    def cmdloop(self, sid=''):
        intro = "Type 'help' for twerp commands"
        self.sid = sid
        if self.sid is not None:
            self.prompt = "twerp (%s) >> " % self.sid[0:7]
        else:
            self.prompt = "twerp (...) >> "
        try:
            return cmd.Cmd.cmdloop(self, sid)
        except KeyboardInterrupt:
            sys.exit(0)

    def do_info(self, sid=None):
        '''
        Display info for current SID
        '''
        if not sid:
            sid = self.sid
        if sid:
            self.client.sid_call(sid)
        else:
            self.logger.error("Need an SID")

    def do_hangup(self, arg=None):
        '''
        Hangs up call associated with current SID shown in prompt.
        '''
        self.client.hangup(self.sid)
        self.prompt = "twerp (...) >> "

    def do_nuke(self, arg=None):
        '''
        Hangs up all voice calls in progress for the entire account
        CAUTION: Read the above carefully.
        '''
        self.client.hangup_all_calls()
        self.prompt = "twerp (...) >> "

    def do_list(self, args=None):
        '''
        List all calls in progress for your account
        '''
        self.client.list_calls()

    def do_sid(self, sid):
        '''
        Change SID for interactive session. New SID shows in prompt.
        '''
        if sid:
            self.sid = sid
            self.prompt = colored.red("twerp (%s...) >> " % self.sid[0:7])
        else:
            print >> sys.stderr, "You must specify the SID to use."

    def do_forward(self, number):
        '''
        Forward call to another phone number. e.g.
        forward +13235551212
        '''
        if number:
            url = '''http://twimlets.com/forward?PhoneNumber='''
            text = quote_plus(number)
            url = '%s%s' % (url, number)
            print url
            call = self.client.call_url(self.sid, url)
            #print >> sys.stderr, call.status

    def do_url(self, url):
        '''
        Redirect call using TwiML at URL e.g.
        url http://twimlets.com/some.twml
        '''
        if url:
            call = self.client.call_url(self.sid, url)
            #print >> sys.stderr, call.status
        else:
            print >> sys.stderr, "You need to specify a valid TWML URL e.g. "
            print >> sys.stderr, \
                "http://twimlets.com/holdmusic?Bucket=com.twilio.music.ambient"

    def do_EOF(self, line):
        '''Exit twerp's interactive mode'''
        return True


class Twerp(object):

    """
    Main class for twerp optparse CLI
    """

    def __init__(self):
        self.options = None
        self.client = RestClient()

    def set_log_level(self):
        """
        Set log level according to command-line options

        @returns: logger object
        """
        log_format = '%(message)s'

        logging.basicConfig(level=logging.ERROR, format=log_format)
        logger = logging.getLogger("twerp")

        if self.options.debug:
            logger.setLevel(logging.DEBUG)
        elif self.options.verbose:
            logger.setLevel(logging.INFO)
        return logger

    def run(self):
        """
        Perform actions based on CLI options

        @returns: status code
        """
        opt_parser = setup_opt_parser()
        (self.options, remaining_args) = opt_parser.parse_args()
        self.logger = self.set_log_level()

        if self.options.numbers:
            return self.client.list_numbers(self.options.verbose)
        elif self.options.dial:
            if not self.options.url and not self.options.say:
                self.logger.error("You must specify a --url with TwiML " +
                    "or --say something")
                return 1
            numbers = self.options.dial.split(",")
            sid = self.client.call_numbers(numbers, self.options.verbose,
                    self.options.callerid, self.options.url,
                    self.options.say)
            if self.options.interactive:
                Interactive(self.client).cmdloop(sid)
                return
        elif self.options.conference:
            if not self.options.room:
                self.logger.error("You must specify a --room to join.")
                return 1
            numbers = self.options.conference.split(",")
            sid = self.client.create_conference(numbers,
                    self.options.room,
                    self.options.verbose,
                    self.options.callerid)
            if self.options.interactive:
                Interactive(self.client).cmdloop(sid)
            return
        elif self.options.purchase:
            return self.client.purchase_number(self.options.purchase)
        elif self.options.contains and self.options.areacode:
            return self.client.numbers_contain_areacode(
                    self.options.areacode,
                    self.options.contains)
        elif self.options.contains:
            return self.client.numbers_contain(self.options.contains)
        elif self.options.areacode:
            return self.client.search_numbers(self.options.areacode)
        elif self.options.rename_account:
             self.client.rename_account(self.options.rename_account)
        elif self.options.list_accounts:
             self.client.list_accounts()
        elif self.options.rename_account:
             self.client.create_subaccount(self.options.create_subaccount)
        elif self.options.applications:
            return self.client.list_applications()
        elif self.options.participants:
            return self.client.list_conference_participants()
        elif self.options.conferences:
            return self.client.list_conferences()
        elif self.options.listsms:
            return self.client.list_sms()
        elif self.options.twerp_version:
            return self.twerp_version()
        elif self.options.sid:
            return self.client.get_sms_sid(self.options.sid)
        elif self.options.sms_recipients:
            numbers = self.options.sms_recipients.split(",")
            return self.client.send_sms(numbers, self.options.sms_message,
                    self.options.verbose)
        elif self.options.notifications:
            return self.client.notifications(self.options.verbose)
        elif self.options.interactive:
            Interactive(self.client).cmdloop('')
            return
        else:
            opt_parser.print_help()
        return 1

    def twerp_version(self):
        """
        Show twerp's version

        @returns: None
        """
        print("twerp version %s" % VERSION)


def setup_opt_parser():
    """
    Setup the optparser

    @returns: opt_parser.OptionParser

    """

    usage = "usage: %prog [options]"
    opt_parser = optparse.OptionParser(usage=usage)

    opt_parser.add_option("--version", action="store_true",
            dest="twerp_version", default=False,
            help="Show twerp version and exit.")

    opt_parser.add_option("-v", "--verbose", action='store_true',
            dest="verbose", default=False, help="Show more output stuff.")

    opt_parser.add_option("--debug", action='store_true',
            dest="debug", default=False, help="Show debugging information.")

    opt_parser.add_option("-q", "--quiet", action='store_true',
            dest="quiet", default=False, help="Show less output.")

    #Common
    group_common = optparse.OptionGroup(opt_parser,
            "Common options",
            "These options can be used for both SMS and voice calls.")

    group_common.add_option("-c", "--callerid", action='store',
            dest="callerid", default=None,
            help="Phone number you are calling from or texting from.")

    group_common.add_option("-i", "--interactive", action='store_true',
            help="Go into interactive command-line mode after " +
            "dialing (voice or conferences).",
            dest="interactive", default=False)

    #SMS
    group_sms = optparse.OptionGroup(opt_parser,
            "SMS options",
            "Send and reveive SMS text messages.")

    group_sms.add_option("-m", "--message", action='store',
            metavar="<TXT MSG>",
            dest="sms_message", default=False, help="Send SMS text message")

    group_sms.add_option("-s", "--sms", action='store',
            dest="sms_recipients", default=False,
            metavar="+12135551212,+14155551212",
            help="Send SMS text message to list of numbers.")

    group_sms.add_option("-l", "--list-sms", action='store_true',
            dest="listsms", default=False, help="Show incoming SMS messages.")

    #Calls (voice)
    group_call = optparse.OptionGroup(opt_parser,
            "Voice call options",
            "Place phone calls, execute TWIML.")

    group_call.add_option("-d", "--dial", action='store',
            metavar="+12135551212,+14155551212",
            help="List of numbers to dial, comma-separated.",
            dest="dial", default=False)

    group_call.add_option("-y", "--say", action='store',
            metavar="Say something.",
            dest="say", default=False,
            help="Use with --dial to say something.")

    group_call.add_option("-u", "--url", action='store',
            metavar="URL of TWIML",
            dest="url", default=False,
            help="URL of TWIML to pass call with --call")

    group_call.add_option("-b", "--buy", action='store',
            dest="purchase", default=False,
            metavar="+12135551212",
            help="Buy a specific phone number listed with -x or -a")

    group_call.add_option("-a", "--area-code", action='store',
            dest="areacode", default=False,
            metavar="AREA CODE",
            help="Search for phone number to purchase by area code. " +
            "Use -b to purchase from these results.")

    group_call.add_option("-x", "--contains", action='store',
            dest="contains", default=False,
            help="Search for phone number to purchase by numbers " +
            "or letters it contains.")

    #Conferences
    group_conferences = optparse.OptionGroup(opt_parser,
            "Conference (voice) options",
            "These options can be used for voice conference calls.")

    group_conferences.add_option("-f", "--conference", action='store',
            metavar="+12135551212,+14155551212",
            help="Start conference with list of numbers to dial, " +
            "comma-separated.",
            dest="conference", default=False)

    group_conferences.add_option("-o", "--room",
            action='store',
            dest="room", default=False,
            help="Room to join for voice conference.")

    group_conferences.add_option("-e", "--conferences", action='store_true',
            dest="conferences", default=False,
            help="Show conferences in-progress.")

    group_conferences.add_option("-p", "--conference-participants",
            action='store_true',
            dest="participants", default=False,
            help="Show participants for all conferences in-progress.")

    #Reports
    group_reports = optparse.OptionGroup(opt_parser,
            "Reporting options",
            "List your Twilio phone numbers and information about each.")

    group_reports.add_option("-n", "--notifications", action='store_true',
            dest="notifications", default=False, help="Show notifications " +
            "from Twilio API (error messages and warnings).")

    group_reports.add_option("-r", "--numbers", action='store_true',
            dest="numbers", default=False, help="Show all my Twilio phone " +
            "numbers. Use -rv for detailed info on each number.")

    group_reports.add_option("--sid", action='store',
            dest="sid", default=False, help="Show log for given SID")

    #Applications
    group_applications = optparse.OptionGroup(opt_parser,
            "Applications",
            "Twilio Application information.")

    group_applications.add_option("--applications", action='store_true',
            dest="applications", default=False,
            help="Show all my Twilio Applications.")

    #Accounts
    group_accounts = optparse.OptionGroup(opt_parser,
            "Accounts options",
            "Twilio account and sub-account management")

    group_accounts.add_option("--list-accounts", action='store_true',
            dest="list_accounts",
            default=False,
            help="List all Twilio accounts and sub-accounts.")

    group_accounts.add_option("--create-sub-account", action='store',
            dest="create_subaccount",
            metavar="NAME",
            default=False,
            help="Create sub-account named 'NAME'")

    group_accounts.add_option("--rename-sub-account", action='store',
            dest="rename_account",
            metavar="NAME",
            default=False,
            help="Rename account or sub-account using 'NAME'")

    opt_parser.add_option_group(group_common)
    opt_parser.add_option_group(group_call)
    opt_parser.add_option_group(group_conferences)
    opt_parser.add_option_group(group_sms)
    opt_parser.add_option_group(group_reports)
    opt_parser.add_option_group(group_applications)
    opt_parser.add_option_group(group_accounts)
    return opt_parser


def main():
    """
    Let's do it.
    """
    my_twerp = Twerp()
    return my_twerp.run()

if __name__ == "__main__":
    sys.exit(main())
